src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -826,48 +834,103 @@ toJSONPBMessageInstD _ctxt parentIdent msgIdent messageParts = do
               , HsVar (unqual_ varNm)
               ]
 
-  -- E.g.
-  -- case f4_or_f9 of
-  --   Just (SomethingPickOneName f4)
-  --     -> HsJSONPB.pair "name" f4
-  --   Just (SomethingPickOneSomeid f9)
-  --     -> HsJSONPB.pair "someid" f9
-  --   Nothing
-  --     -> mempty
-  let oneofCaseE (OneofField{subfields}) =
-        HsCase disjunctName (altEs <> pure fallThruE)
+  -- Note: [Generating eta-reduced code]
+  -- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  -- Suppose we have a sum type Foo, nested inside a message Bar.
+  -- We want to generate the following:
+  --
+  -- > toJSONPB (Bar foo more stuff) =
+  -- >   HsJSONPB.object
+  -- >     [ (\caseExpr option -> if optEmitInlinedOneof option
+  -- >                            then oneofCaseE option
+  -- >                            else PB.object ["Foo" .= (PB.object [caseExpr] option)] option
+  -- >       ) (<case expr scrutinising foo> :: Options -> Value)
+  -- >     , <do more>
+  -- >     , <do stuff>
+  -- >     ]
+  --
+  -- To avoid gensymming parameters for the lambda, we will instead generate the
+  -- following pointfree code.
+  --
+  -- > liftA3 bool
+  -- >   <$> ((flip ("Foo" .=) <*>) . PB.object . pure)
+  -- >   <*> id
+  -- >   <*> pure optEmitInlinedOneof
+  --
+  -- ==>
+  --
+  -- > (liftA3 (liftA3 bool))
+  -- >   ((flip ("Foo" .=) <*>) . PB.object . pure)
+  -- >   id
+  -- >   (pure optEmitInlinedOneof)
+  --
+  -- The interested reader is invited to verify that this transformation is
```

Comment:
I know you don't mean it this way but I read this with a patronizing voice. I know this is standard fare in textbooks but I think we can do better in our opensource code by providing a test to show that it is a correct transformation (if you haven't already); thus, also enlightening the reader.

Reply:
I also think that if there aren't any clear benefits to the eta-reduced code, then code that _does not require_ the reader to go through the exercise of program calculation (or following a test verifying the transformation) is better and clearer.

Reply:
I want to further stress the importance of making the generated code easier to read by quoting this:

> Everyone knows that debugging is twice as hard as writing a program in the first place. So if you're as clever as you can be when you write it, how will you ever debug it?
>
> - Brian Kernighan

Reply:
I still believe the generated code should be easy for people to read.  Also, all free variables within the example are statically specified by the code generator so I believe there would be no risk of a free variable being shadowed by our bound variable named `options`